lib/Commands/NinjaBuildCommand.cpp

```diff
@@ -479,38 +480,20 @@ class BuildContext {
   /// The limited queue we use to execute parallel jobs.
   std::unique_ptr<BuildExecutionQueue> jobQueue;
 
-  /// The previous SIGINT handler.
-  struct sigaction previousSigintHandler;
-
   /// The set of spawned processes to cancel when interrupted.
   std::unordered_set<pid_t> spawnedProcesses;
   std::mutex spawnedProcessesMutex;
-
-  /// Low-level flag for when a SIGINT has been received.
-  static std::atomic<bool> wasInterrupted;
-
-  /// Pipe used to allow detection of signals.
-  static int signalWatchingPipe[2];
-
-  static void sigintHandler(int) {
-    // Set the atomic interrupt flag.
-    BuildContext::wasInterrupted = true;
-
-    // Write to wake up the signal monitoring thread.
-    char byte{};
-    sys::write(signalWatchingPipe[1], &byte, 1);
-  }
-
+  
   /// Cancel the build in response to an interrupt event.
-  void cancelBuildOnInterrupt() {
-    std::lock_guard<std::mutex> guard(spawnedProcessesMutex);
+  static void interruptedHandler(BuildContext *context) {
+    std::lock_guard<std::mutex> guard(context->spawnedProcessesMutex);
```

Comment:
Actually your current implementation does that because it uses `std::function`, you could register a lambda for the handler already.